src/core/datatypes/trees/TopologyNode.cpp

```diff
@@ -1817,21 +1817,9 @@ void TopologyNode::renameNodeParameter(const std::string &old_name, const std::s
 
 void TopologyNode::setAge(double a, bool propagate)
 {
-    // Sometimes the `sampled_ancestor` flag is set to `true` for nodes with 1 child.
-    // For those nodes we want to set the age directly, not modify the parent.
-    if ( isTipSampledAncestor() and propagate == true )
+    if ( sampled_ancestor == true && propagate == true )
     {
```

Comment:
I don't understand what the propagate argument is used for here
also I don't think we should silently change the age of SA, considering this is what triggered the whole issue ? my proposal would be to have an argument `allow_SA_changes` that is by default false - if argument is true, proceed otherwise throw an exception
then we can set it to true only for things that are explicitly allowed to change taxon ages (i.e. only `TipTimeSlideProposal` at the moment)
bonus - it will tell us immediately which things are trying to change SA ages without being allowed to

Reply:
This is a revert to the previous state to make sure that I didn't break it.  Now that it is clear that I didn't break this, I think we can go ahead with a real fix, although I guess we still have to make sure not to break anything in the process.

Sorry, when I first messaged you I thought that this changed fixed the problem, but then I realized that it did not.

The `propagate` argument seems to mean that when modifying the age of a sampled ancestor tip, one should change the parent as well.

1. One option is to modify setAge so that changing the age of the sampled-ancestor-parent always aborts.  The rule would be that if you want to modify the and of a sampled-ancestor, you have to always modify the age of the tip node.

2. But I kind of like your idea of allowing changes via the sampled-ancestor-parent if `allow_SA_changes` is true.

Another idea was to make a new routine called proposeAge that is a wrapper around setAge that is used in MCMC moves and proposals.

Reply:
Also: why is it so important that only mvTipTimeSlideUniform changes the node ages?

Reply:
so in practice I think analyses with fossil samples which have a fixed age or a narrow range of ages could become very inefficient if moves are allowed to change node ages without caring about the type of node, but then rejected when/if it doesn't work with a SA
I am also trying to prevent unintended broken setups - right now it's very clear from a user perspective whether they intend to estimate tip ages (in which case they include the mvTipTimeSlideUniform) or not, and I think it would be good to keep it that way
and it's also a good early warning for someone coding a new move who hasn't thought about that issue and maybe should
of course it could be that I'm just trying to replicate what I'm used to in BEAST2, so take that with a grain of salt :)

Reply:
- mvTipTimeSlideUniform already looks at the age constraints for a node in order to avoid inefficiency.  We could replicate that in mvNodeTimeSlideUniform.  In theory we could even combine the two moves.
-  I don't want to break user setups, and I think it should be clear whether tip ages are fixed.  But I don't think it should be based on whether or not an MCMC move is included. 
- Can we specify whether tip ages are fixed by how we read taxon data?  We could split readTaxonData( ) into readTipAges( ) and readTipAgeRanges( ), for example.
- I used to like the early warning idea, but now I feel more conflicted about it.
- I definitely do not want to replicate BEAST2 without a good reason :-P

Reply:
- I was more thinking of tree moves such as tree scaling, tree rearrangements etc which may need special handling for SAs nodes
- my objection to splitting readTaxonData is that it's not going to be backwards-compatible with previous analyses, and we have already enough issues open due to function name changes :p